.codeqlmanifest.json

```diff
@@ -1,11 +1,16 @@
-{ "provide": [ "ruby/.codeqlmanifest.json",
-                "*/ql/src/qlpack.yml",
-               "*/ql/lib/qlpack.yml",
-               "*/ql/test/qlpack.yml",
-               "cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/tainted/qlpack.yml",
-               "*/ql/examples/qlpack.yml",
-               "*/upgrades/qlpack.yml",
-               "javascript/ql/experimental/adaptivethreatmodeling/lib/qlpack.yml",
-               "javascript/ql/experimental/adaptivethreatmodeling/src/qlpack.yml",
-               "misc/legacy-support/*/qlpack.yml",
-               "misc/suite-helpers/qlpack.yml" ] }
+{
+    "provide": [
+        "*/ql/src/qlpack.yml",
+        "*/ql/lib/qlpack.yml",
+        "*/ql/test/qlpack.yml",
+        "cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/tainted/qlpack.yml",
+        "*/ql/examples/qlpack.yml",
+        "*/upgrades/qlpack.yml",
+        "javascript/ql/experimental/adaptivethreatmodeling/lib/qlpack.yml",
+        "javascript/ql/experimental/adaptivethreatmodeling/src/qlpack.yml",
+        "misc/legacy-support/*/qlpack.yml",
+        "misc/suite-helpers/qlpack.yml",
+        "ruby/extractor-pack/codeql-extractor.yml",
```

Comment:
@dbartol I'm fine with this change, but I was wondering what prompted you to make it? It is to make things more consistent, or did you need to change this to workaround some technical limitation/complication?

Personally, I'd rather see language specific stuff in per-language manifest files and keep the top-level manifest clean. Of course doing things inconsistently like we had before is worse, so we should either move:
```
        "cpp/ql/test/query-tests/Security/CWE/CWE-190/semmle/tainted/qlpack.yml",
        "javascript/ql/experimental/adaptivethreatmodeling/lib/qlpack.yml",
        "javascript/ql/experimental/adaptivethreatmodeling/src/qlpack.yml",
        "ruby/extractor-pack/codeql-extractor.yml",
        "ruby/extractor-pack/codeql-extractor.yml",
```
out of this file, or keep them all in the top-level file like you did. I have a slight preference for moving things out, but I leave it up to you to decide (or flip a coin ;-) )

Reply:
The `.codeqlmanifest.json` (or its modern equivalent, `codeql-workspace.yml`) file defines a "CodeQL Workspace", which includes settings that apply to all packs directly contained in that workspace. That's where the set of available change note categories are defined, the versioning policies, and so forth. If we leave Ruby in its own, it would need its own copies of these policies.

Also, pack dependencies are automatically resolved when the target pack is in the same workspace, so with a separate `.codeqlmanifest.json` for Ruby, the `codeql/suite-helpers` pack can't be resolved by default unless you run `codeql pack install` to download a published version, or add a `--additional-packs` path to the command line.

Reply:
I'm not sure I completely follow what you say. The `.codeqlmanifest.json` file existed before the concept of "CodeQL Workspace" was invented, right? I suspect the semantics of `.codeqlmanifest.json`  has been retro-fitted (and changed) to match the idea of a "CodeQL Workspace".

Isn't the problem here is that you don't seem to allow "nested" workspaces? A bit like maven's idea of a "parent" pom file? No need to implement nested workspaces, the simple solution of having only one large workspace is fine with me. I'm just trying to understand things.

Reply:
I prefer to think of it as "`.codeqlmanifest.json` has evolved and expanded to become a 'workspace'". Originally, `.codeqlmanifest.json` was used to allow you to add a single directory to your search path, but have the search look in many different paths under that directory. That aspect of it still applies. However, we now automatically resolve references between packs within the same workspace. Without Ruby being in the same workspace as the `suite-helpers` pack, you'd need to specify a search path. Before workspaces, you would have always needed a search path.

We currently support nested workspace in a somewhat limited sense: A pack in a parent workspace that references a pack in a nested workspace will automatically find the pack in that nested workspace. However, a pack in a nested workspace will not find packs in the parent workspace, nor in sibling workspaces. An early implementation of workspaces actually had this feature, but it was removed during review because we didn't actually have a real-world use case for it at the time. I'm open to adding that feature back, but I'm not sure I'd do it just to make one of our own workspaces only slightly cleaner.

Reply:
Thanks for the explanation!